operator/pkg/controllers/etcd/etcd.go

```diff
@@ -0,0 +1,346 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	etcdRootCACommonName = "etcd/ca"
+)
+
+type Provider struct {
+	kubeClient client.Client
+}
+
+type CertConfig struct {
+	*certutil.Config
+	name   string
+	isCA   bool
+	caCert *x509.Certificate
+	caKey  crypto.Signer
+}
+
+func New(kubeclient client.Client) *Provider {
+	return &Provider{kubeClient: kubeclient}
+}
+
+func (e *Provider) Reconcile(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	// generate etcd service object
+	etcdServiceName, err := e.createETCDService(ctx, controlPlane)
+	if err != nil {
+		return err
+	}
+	// Create ETCD certs and keys, store them as secret in the management server
+	if err := e.createETCDCerts(ctx, controlPlane); err != nil {
+		return err
+	}
+	// TODO generate etcd stateful set
+	_ = etcdServiceName
+	return nil
+}
+
+func (e *Provider) createETCDService(ctx context.Context, controlPlane *v1alpha1.ControlPlane) (string, error) {
+	objName := etcdServiceNameFor(controlPlane.ClusterName())
+	service, err := e.getService(ctx, namespacedName(controlPlane.NamespaceName(), objName))
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if err := e.createService(ctx, controlPlane); err != nil {
+				return "", fmt.Errorf("creating etcd service object for cluster %v, %w", controlPlane.ClusterName(), err)
+			}
+			return objName, nil
+		}
+		return "", fmt.Errorf("getting service from api server, %w", err)
+	}
+	return service.Name, nil
+}
+
+func (e *Provider) createService(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      etcdServiceNameFor(controlPlane.ClusterName()),
+			Namespace: controlPlane.NamespaceName(),
+			Labels:    etcdLabelFor(controlPlane.ClusterName()),
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: controlPlane.APIVersion,
+				Name:       controlPlane.Name,
+				Kind:       controlPlane.Kind,
+				UID:        controlPlane.UID,
+			}},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Port: 2380,
+				Name: etcdServerPortNameFor(controlPlane.ClusterName()),
+			}, {
+				Port: 2379,
+				Name: etcdClientPortNameFor(controlPlane.ClusterName()),
+			}},
+			ClusterIP: "None",
+			Selector:  etcdLabelFor(controlPlane.ClusterName()),
+		},
+	}
+	if err := e.kubeClient.Create(ctx, svc); err != nil {
+		return fmt.Errorf("creating service %s/%s, %w", svc.Namespace, svc.Name, err)
+	}
+	zap.S().Infof("[%s] created service %s", controlPlane.ClusterName(),
+		etcdServiceNameFor(controlPlane.ClusterName()))
+	return nil
+}
+
+func (e *Provider) getService(ctx context.Context, name types.NamespacedName) (*v1.Service, error) {
+	result := &v1.Service{}
+	if err := e.kubeClient.Get(ctx, name, result); err != nil {
+		return nil, fmt.Errorf("getting service %v, %w", name, err)
+	}
+	return result, nil
+}
+
+func (e *Provider) createETCDCerts(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
```

Comment:
I would probably orient the logic similarly with separating construction and application of secrets. The construction can also be generic, since we need so much pki for kubernetes.
```
certificateFor(...) // return a secret
// create if not exists
```